Role changes to support enforcing workflow restrictions

slobodanadamovic opened the pull request:
This PR implements necessary changes to `Role` classes in order to enforce 
workflow restrictions for API keys. Main change is around resolving roles from 
role descriptors, where a role can either be effectively empty (or not) 
depending on its workflow restriction. 

Currently, the only workflow supported is `search_application_query` which 
allows restricting access only to [Search Application Search API](https://www.elastic.co/guide/en/elasticsearch/reference/8.8/search-application-search.html).

For simplicity, when creating an API key, it's possible to specify only a single 
role descriptor with workflows restriction. For example:

```
POST /_security/api_key
{
  "name": "my_restricted_api_key",
  "role_descriptors": {
    "my_restricted_role": {
      "indices": [
        {
          "names": ["books"],
          "privileges": ["read"]
        }
      ],
      "restriction": {
        "workflows": ["search_application_query"]
      }
    }
  }
}
```

Relates to https://github.com/elastic/elasticsearch/pull/96215

slobodanadamovic committed:
Role changes to support enforcing workflow restrictions

slobodanadamovic committed:
Change role filtering to explicit workflow restriction checking

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-role-restriction

# Conflicts:
#	x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
#	x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java
#	x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/SecurityRestFilterTests.java

slobodanadamovic committed:
Implement build role from role descriptor for workflows

slobodanadamovic committed:
Add assertions

slobodanadamovic committed:
Add more tests

slobodanadamovic committed:
Change modifier to private

slobodanadamovic committed:
Update assertion

slobodanadamovic committed:
Remove unused parameter

slobodanadamovic committed:
Test restriction with search application query API

slobodanadamovic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/LimitedRole.java:
```diff
@@ -74,6 +78,29 @@ public RemoteIndicesPermission remoteIndices() {
+            return limitedByRole.workflowsRestriction();
+        }
+        if (false == limitedByRole.hasWorkflowsRestriction()) {
+            return baseRole.workflowsRestriction();
```
The implementation of this method can be simplified to just return `baseRole.workflowsRestriction()` since we don't expect that `limitedByRole` has any workflow restriction.

slobodanadamovic committed:
Implement role filtering approach

slobodanadamovic committed:
name nit

slobodanadamovic committed:
Apply spotless

slobodanadamovic committed:
Simplify role filtering

slobodanadamovic committed:
Fail early - immediately after authz info resolution

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:
```diff
@@ -429,11 +438,17 @@ public static void buildRoleFromDescriptors(
             }
+
+            if (descriptor.hasWorkflowsRestriction()) {
+                assert roleDescriptors.size() == 1 : "only single role descriptor with workflow restriction is allowed";
```
I think this should be an exception to counter potential abuse of custom role provider

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/Role.java:
```diff
@@ -259,6 +271,11 @@ public Builder addApplicationPrivilege(ApplicationPrivilege privilege, Set<Strin
         }
 
+        public Builder workflows(Set<String> workflowNames) {
+            this.workflowsRestriction = WorkflowsRestriction.resolve(workflowNames);
```
We don't need to resolve Workflow again here. WorkflowsRestriction can really just be a StringMatcher internally, something like
```
WorkflowsRestriction(Set<String> names) {
    this.names = names;
    this.predicate = StringMatcher.of(names);
}
```

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/restriction/WorkflowsRestriction.java:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * Default behaviour is to allow access to all workflows if none are provided.
+     */
+    public static final WorkflowsRestriction ALLOW_ALL = new WorkflowsRestriction(Set.of());
```
I'd suggest rename this to `NONE` since this actually means no restriction at all, i.e. nothing has been configured.

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/restriction/WorkflowsRestriction.java:
```diff
@@ -0,0 +1,50 @@
+        return this.workflows.size() > 0;
+    }
+
+    public boolean isWorkflowAllowed(Workflow workflow) {
```
Similar to my other comment, I suggest we just work with `String` at this layer.

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/LimitedRole.java:
```diff
@@ -74,6 +76,22 @@ public RemoteIndicesPermission remoteIndices() {
+        if (baseRestricted == EMPTY_RESTRICTED_BY_WORKFLOW) {
+            return EMPTY_RESTRICTED_BY_WORKFLOW;
+        }
+        // We don't support workflows restriction for limitedByRole,
+        // hence it is safe to return this object unchanged.
+        return this;
```
I don't know why it is safe to return unchanged? Checking whether limitedByRole is empty and restricted may be redundant but feels technically "safer" to me. Also the `Role.limitedBy(..)` does not prevent it so I think we should be consistent.

slobodanadamovic replied:
You are right. It's safer to add an additional check instead of to assume.

slobodanadamovic committed:
Check limitedByRole if it's restricted. Change assert to IAE

slobodanadamovic committed:
Address review feedback

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:
```diff
@@ -142,10 +142,13 @@ public RBACEngine(
-            listener.delegateFailureAndWrap((l, roleTuple) -> l.onResponse(new RBACAuthorizationInfo(roleTuple.v1(), roleTuple.v2())))
-        );
+        rolesStore.getRoles(authentication, listener.delegateFailureAndWrap((l, roleTuple) -> {
+            if (roleTuple.v1() == Role.EMPTY_RESTRICTED_BY_WORKFLOW || roleTuple.v2() == Role.EMPTY_RESTRICTED_BY_WORKFLOW) {
+                l.onResponse(DeniedAuthorizationInfo.INSTANCE);
```
I don't think this is an issue, but would like to write down for prosperity: `v1` is the run-as user's role, `v2` is the authenticating user's role. If `v1` is empty and restricted while `v2` is a proper role, one might argue that `v2` should be able to run-as `v1` (i.e. pass `authorizeRunAs`) then fail. 

I think it can be argued both ways because:

We have a similar situation where request fails before authorization happens if [run-as user is disabled](https://github.com/elastic/elasticsearch/blob/31a47867428f1f53bfa72b049fd5e082a4fef234/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java#L320-L321). 

At the same time, we also have request fails at authorization if the [run-as user does not exist](https://github.com/elastic/elasticsearch/blob/31a47867428f1f53bfa72b049fd5e082a4fef234/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java#L222-L223). 

These two cases do not agree with each other. But more often we view the later one as an issue since it make an exception that `realm` can be `null` and has been causing issues in handling it. Therefore I overall feel it's OK to fail earlier here. Letting it passing through is not actually useful other than for argument sake.

The practical bottom line is that it is not possible to run-as API keys. So not a real issue anyway.

slobodanadamovic replied:
Thanks for raising this topic. FWIW My reasoning was that both users (authenticating and run-as) should have access to a particular workflow.  Allowing the authenticating user to invoke a workflow should be prevented early on, even if run-as user can access it.

Additionally, what we could potentially do is differentiate between which role was restricted, but this felt unnecessary at the moment.

ywangd replied:
> Allowing the authenticating user to invoke a workflow should be prevented early on, even if run-as user can access it.

My view is the opposite. When
1. The API is a workflow API 
2. The authenticating user has the privilege to run-as the effective user
3. The effective user has a workflow role that allows it to call the workflow API

In this situation, the authorization should succeed. I believe that's what naturally would happen with the proposed change. Because Item 2 implies the authenticating user must have access to the workflow as well, either because the authenticating user has a workflow role or has a role _without_ workflow restriction. 

My original comment was about the failure case. That is, Item 1 & 2 remain unchanged, but in Item 3, the effective user does *not* have access to the workflow API. In this case, we could let the run-as (Item 2) to be successful (it generates an audit log entry `runAsGranted`), then fail due to Item 3. Or we can fail before even check Item 2 (and skip the `runAsGranted` audit entry), which is the behaviour of the proposed change. Both options are valid to some extend. But going with the later option seems to give us more flexibility in future (thinking moving rejection to the REST layer).

slobodanadamovic committed:
Fix failing tests

slobodanadamovic committed:
Fail authz info resolving instead of returning denied object

slobodanadamovic committed:
Revert accidentally removed check

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-role-restriction

# Conflicts:
#	server/src/main/java/org/elasticsearch/TransportVersion.java

slobodanadamovic committed:
Change assert to IAE

slobodanadamovic committed:
Fix failing test.

slobodanadamovic committed:
Change back to assertion

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-role-restriction

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-role-restriction

# Conflicts:
#	server/src/main/java/org/elasticsearch/TransportVersion.java

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

slobodanadamovic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:
```diff
@@ -316,13 +317,18 @@ public void authorize(
+                    maybeAuthorizeRunAs(requestInfo, auditId, authorizationInfo, listener);
+                }, e -> {
+                    if (e instanceof ElasticsearchAccessRestrictedException) {
+                        auditTrailService.get()
```
Based on our discussion (and also for completeness), I've added access denied audit log. We can re-evaluate this decision and most probably introduce dedicated audit even in a follow-up work when we move the role resolution earlier in the authorization flow.

slobodanadamovic committed:
Fix condition check

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:
```diff
@@ -142,10 +143,13 @@ public RBACEngine(
-            authentication,
-            listener.delegateFailureAndWrap((l, roleTuple) -> l.onResponse(new RBACAuthorizationInfo(roleTuple.v1(), roleTuple.v2())))
-        );
+        rolesStore.getRoles(authentication, listener.delegateFailureAndWrap((l, roleTuple) -> {
+            if (roleTuple.v1() == Role.EMPTY_RESTRICTED_BY_WORKFLOW || roleTuple.v2() == Role.EMPTY_RESTRICTED_BY_WORKFLOW) {
+                l.onFailure(new ElasticsearchAccessRestrictedException("access restricted by workflow"));
+            } else {
+                l.onResponse(new RBACAuthorizationInfo(roleTuple.v1(), roleTuple.v2()));
+            }
+        }));
```
It feels a bit weird that the overloading method below
```
resolveAuthorizationInfo(Subject subject, ActionListener<AuthorizationInfo> listener)
```
does nothing for empty-restricted role. 

Looking at its use cases, hasPrivileges, profileHasPrivileges, resolving remote indices for RCS 2.0, I think there might be things that we need dig deeper.

* HasPrivileges - If the code reaches the overloading method, it means the user must have a non-empty-restricted role. So this use case should be fine.
* ProfileHasPrivileges - The API is called by Kibana to check an end-user's privileges. In this case, what should be the workflow for the end-user? It effectively will always behave as no incoming-workflow because ProfileHasPrivileges API does not belong to any workflow. This might not be correct in all cases. But API keys cannot have profiles so it is not an issue today.
* Resolving remote indices - This has two sub-use-cases. 
  * The first one is ad-hoc query and this should be fine because when the code reaches here, we can already be sure the user has a non-empty-restriced role. 
  * The second one is a persistent task which serializes user's authentication object as part of the job definition. Since workflow is another header in threadContext, it should also be serialized for the job. Now this becomes trickly, the serialized workflow would be the API that creates the persistent task, e.g. for transform jobs, this is the PutTransform API. But when the task runs, it runs a search request which would corresponds to a different REST API. In this case, how should workflow work? 

I am not quite sure yet nhow all these could be rationalized into a consistent behaviour. Fortunately, none of them is likely a pratical problem today. So we can revisit it later. 

For now, I suggest we put a TODO comment in the below method and explain briefly why we are leaving it out in this PR.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/restriction/WorkflowService.java:
```diff
@@ -33,12 +33,12 @@ public Workflow resolveWorkflowAndStoreInThreadContext(RestHandler restHandler,
-    public Workflow readWorkflowFromThreadContext(ThreadContext threadContext) {
+    public String readWorkflowFromThreadContext(ThreadContext threadContext) {
         String workflowName = threadContext.getHeader(WORKFLOW_HEADER);
         if (workflowName == null) {
             return null;
         }
-        return WorkflowResolver.resolveWorkflowByName(workflowName);
+        return workflowName;
```
Nit: I think this can be simplified as just
```
return threadContext.getHeader(WORKFLOW_HEADER);
```

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/Role.java:
```diff
@@ -58,6 +61,10 @@ public interface Role {
 
+    boolean hasWorkflowsRestriction();
+
+    Role forWorkflow(String workflow);
```
Nit: can we add `@Nullable` to `String workflow` since it is not obvious that it can be null.

slobodanadamovic replied:
Sure, I've also added javadoc.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:
```diff
@@ -429,11 +437,19 @@ public static void buildRoleFromDescriptors(
                 });
             }
+
+            if (descriptor.hasWorkflowsRestriction()) {
+                if (roleDescriptors.size() != 1) {
+                    throw new IllegalArgumentException("only single role descriptor with workflow restriction is allowed");
+                }
```
I suggest we check this in the beginning of the method (right after checking whether list of role descriptor is empty) and call `listener.onFailure` if there is violation (instead of throwing). Also nit: can we add the role descriptor names in the error message?

slobodanadamovic replied:
Thanks for the suggestion. I've implemented the [validation](https://github.com/elastic/elasticsearch/pull/96744/files#diff-65dc248b6cd87fca037bd9e993b2aab3828578878ab92e507a88ce89925b493eR385-R408) and [added tests](https://github.com/elastic/elasticsearch/pull/96744/files#diff-57010f0ef921acba217dce681c900c49393738fd80e79f0709cb2a241ea1a558R1093-R1118) for it.

ywangd reviewed (commented):
I still need read test code. Here are some comments for the main code changes. Thanks!

slobodanadamovic committed:
Address review fedback

slobodanadamovic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/restriction/WorkflowsRestriction.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Default behaviour is to allow access to all workflows if none are provided.
+     */
+    public static final WorkflowsRestriction NONE = new WorkflowsRestriction(Set.of());
```
Good point. Makes sense to keep it consistent with RoleDescriptor. I've adjusted it as suggested.

slobodanadamovic replied:
This suggestion also allowed me to simplify the checks in `isWorkflowAllowed`.

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-role-restriction

# Conflicts:
#	server/src/main/java/org/elasticsearch/TransportVersion.java

slobodanadamovic committed:
Fix failing test

slobodanadamovic committed:
Remove unnecessary resolve method. Update javadoc.

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-role-restriction

slobodanadamovic committed:
Add a TODO for future reference

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-role-restriction

# Conflicts:
#	server/src/main/java/org/elasticsearch/TransportVersion.java

slobodanadamovic commented:
@elasticmachine run elasticsearch-ci/bwc

ywangd reviewed (approved):
LGTM

I left quite a few comments. I'd appreciate if you could address them. Three of them are of some importance:
* Allow [empty list of workflow names](https://github.com/elastic/elasticsearch/pull/96744#discussion_r1233466748) in `WorkflowRestriction` to mean nothing is permitted (workflow or not)
* Call `Role.forWorkflow(...)` in `CompositeRolesStore.getRole` on the [final role](https://github.com/elastic/elasticsearch/pull/96744#discussion_r1233475965) built from role reference.
* [Audit log](https://github.com/elastic/elasticsearch/pull/96744#discussion_r1233420686) or not

Please ping me if you need any clarifications. Thanks!

In addition, I think we need the `>feature` label instead of `>non-issue` for this PR since the doc PR will likely land after FF. So it is better for another PR to carry the `>feature` label and I think this PR is likely the best candidate.

slobodanadamovic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/Role.java:
```diff
@@ -46,6 +48,8 @@ public interface Role {
 
     Role EMPTY = builder(new RestrictedIndices(Automatons.EMPTY)).build();
 
+    Role EMPTY_RESTRICTED_BY_WORKFLOW = builder(new RestrictedIndices(Automatons.EMPTY)).build();
```
I like this better as well. The only catch is that in `WorkflowsRestriction` constructor we have a special handling for an empty case and `null` workflow check since immutable collections do not allow null values when calling `contains`. So I think, it should look like this:

```
...
if (names == null) {
    // No restriction, all workflows are allowed
    this.predicate = name -> true;
} else if (names.isEmpty()) {
    // Empty restriction, no workflow is allowed
    this.predicate = name -> false;
} else {
    this.predicate = name -> {
        if (name == null) {
            return false;
        } else {
            return names.contains(name);
        }
    };
}
```

ywangd replied:
Cool that looks good to me. 

>  immutable collections do not allow null values when calling contains

Right. StringMatcher actually uses HashSet internally. Thanks!

slobodanadamovic committed:
Address review feedback:

- Check same instance instead equalTo
- Allow empty list of workflow names in WorkflowRestriction to mean nothing is permitted
- Call Role.forWorkflow(...) in CompositeRolesStore.getRole on the final role
- Remove audit logging
- ElasticsearchRoleRestrictionException

slobodanadamovic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:
```diff
@@ -282,7 +288,8 @@ public void buildRoleFromRoleReference(RoleReference roleReference, ActionListen
         } else {
-            roleActionListener.onResponse(existing);
+            String workflow = workflowService.readWorkflowFromThreadContext(threadContext);
+            roleActionListener.onResponse(existing.forWorkflow(workflow));
```
Good catch. I missed this, as I was making changes bottom-up. Hence why I end-up needing a short-circuit in `Role.limitedBy`.

slobodanadamovic committed:
Apply spotless.

slobodanadamovic commented on x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/apikey/ApiKeyWorkflowsRestrictionRestIT.java:
```diff
@@ -0,0 +1,215 @@
+        indexDocRequestA.setJsonEntity("{\"foo\": \"bar\"}");
+        assertOK(adminClient().performRequest(indexDocRequestA));
+
+        final Request indexDocRequestB = new Request("POST", "/index-b/_doc/doc2?refresh=true");
+        indexDocRequestB.setJsonEntity("{\"baz\": \"qux\"}");
+        assertOK(adminClient().performRequest(indexDocRequestB));
```
You are right, it does not add value at all. My plan was to have a separate Search Application for `index-b` only, in order to demonstrate that indices privileges are respected and access is denied even when workflow restriction allows access. I somehow missed to add this test. I'll add it now.

slobodanadamovic committed:
Add tests for CompositeRolesStore.getRole

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-role-restriction

# Conflicts:
#	server/src/main/java/org/elasticsearch/TransportVersion.java

slobodanadamovic committed:
Fix failing tests by spying on real instance to avoid that callRealMethod fails with NPE

slobodanadamovic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/restriction/WorkflowsRestriction.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public boolean isWorkflowAllowed(String workflow) {
+        return predicate.test(workflow);
```
> can we remove StringMatcher and replace it with just Set::contains as the predicate

I've removed it. Makes more sense to use `Set::contains` since we don't support wildcards at all.

slobodanadamovic committed:
Add debug log.

slobodanadamovic committed:
Test RBACEngine.resolveAuthorizationInfo

slobodanadamovic committed:
Test AuthorizationService access denial

slobodanadamovic commented on x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/apikey/ApiKeyWorkflowsRestrictionRestIT.java:
```diff
@@ -0,0 +1,215 @@
+        queryRequest.setJsonEntity("""
+            {
+              "params": {
+                "value": "bar",
+                "size": 10,
+                "from": 0,
```
🤦 Copy-paste mistake of search application example. I've removed them as they are not needed and changed to call `GET _application/search_application/my-app/_search` instead.

slobodanadamovic committed:
Apply review comments to ApiKeyWorkflowsRestrictionRestIT

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-role-restriction

slobodanadamovic commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java:
```diff
@@ -1086,6 +1090,32 @@ public void testBuildRoleWithSingleRemoteIndicesDefinition() {
         assertHasIndexGroupsForClusters(role.remoteIndices(), Set.of(clusterAlias), indexGroup("index-1"));
     }
 
+    public void testBuildRoleFromDescriptorsWithSingleRestriction() {
```
Indeed. Ive added two tests as suggested:
1. `testGetRoleForWorkflowWithoutRestriction`
2. `testGetRoleForWorkflowWithRestriction`

slobodanadamovic committed:
Update docs/changelog/96744.yaml

elasticsearchmachine commented:
Hi @slobodanadamovic, I've created a changelog YAML for you.

slobodanadamovic committed:
Update docs/changelog/96744.yaml
